feat: support mention in Research descriptions, updates and comments

thisislawatts opened the pull request:
PR Checklist

- [X] - Commit [messages are descriptive](https://github.com/ONEARMY/community-platform/blob/master/CONTRIBUTING.md#--commit-style-guide), it will be used in our [Release Notes](https://github.com/ONEARMY/community-platform/releases/)

PR Type

- [ ] Bug fix (non-breaking change which fixes an issue)
- [X] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Developer experience (improves developer workflows for contributing to the project)

## Description

_What this PR does_

## Git Issues

Closes #

## Screenshots/Videos

_If useful, provide screenshot or capture to highlight main changes_

---

## What happens next?

Thanks for the contribution! We try to make sure all PRs are reviewed ahead of a monthly dev call (first Monday of the month, open to all!).

If the PR is working as intended it'll be merged and included in the next platform release, if not changes will be requested and re-reviewed once updated.

If you need more immediate feedback you can try reaching out on Discord in the [Community Platform `development` channel](https://discord.com/channels/586676777334865928/938781727017558018).

davehakkens commented:
Thanks @thisislawatts nice feature!
All seems to work here, mentions are linked + notifications triggered.
Only thing is that the text of the notification is empty. (does work for how-to)
The last four notifications are
1. Mentioned in first post
2. Mentioned in a step
3. Mentioned in research comment
4. Mentioned in research comment
<img width="258" alt="afbeelding" src="https://user-images.githubusercontent.com/13672737/214969155-7d757807-3d9a-42cc-a704-ea9fd05b387a.png">

thisislawatts commented:
@davehakkens what should the wording be for these notifications. Do we want to specify whether a user was mentioned in a Research description, Update or comment?

davehakkens commented:
We could make it very basic and say for all mentions (content + comment) the same:
 _[[username]](link-to-profile.com) mentioned you [here](link-to-content.com)_
Direct to wherever the mention is made.

However you already made some better/ more advanced behaviour in the howto mentions. Separating a comment from the content.
- Mentioned in a how-to
- Mentioned in a comment

So I would continue like this and do the same for research
- Mentioned in a research (description + update)
- Mentioned in a comment

thisislawatts commented:
When testing this manually with a URL like: http://localhost:3000/research/testing-mentions#update-0-comment-2, there is a UX problem here. The user has been mentioned in the 3rd comment under the 1st update. Whilst that information is present in the document fragment `#update-0-comment-2` the page does not reflect that. The comments are collapsed and hidden from view.

thisislawatts commented:
@davehakkens have you seen notifications that match the format you've described here? 

> However you already made some better/ more advanced behaviour in the howto mentions. Separating a comment from the content.
> 
> - Mentioned in a how-to
> - Mentioned in a comment

Looking at the existing code I can't see anything which suggests the system is capable of generating them. This is the current implementation:

```tsx
case 'howto_mention':
  return (
    <Box>
      You were mentioned in a
      <InternalLink to={notification.relevantUrl || ''}>
        how-to
      </InternalLink>
      by
      <InternalLink to={'/u/' + notification.triggeredBy.userId}>
        {notification.triggeredBy.displayName}
      </InternalLink>
    </Box>
 )
```

I agree that it would be good to distinguish on them.

thisislawatts commented:
Ready for a re-review ✨

davehakkens commented:
> @davehakkens have you seen notifications that match the format you've described here?
> 
Pretty sure I didnt. but cant seem to find it anymore either 🤔
Perhaps I was mixing these two notifications:

**For how-to author:** 
New comment on your [how-to](http://localhost:3000/how-to/test-title-change-2) by [projectkamp](http://localhost:3000/u/5Q1yqCxLgOOBMonR3kjJv5Od3EQ2)

**For being mentioned in how-to comment:** 
You were mentioned in a [how-to](http://localhost:3000/how-to/test-title-change-2#comment:eduxW6BS3DTi7XLj1GvG) by [projectkamp](http://localhost:3000/u/y3mcw6P2lmNPQS6Fi9e7uSpvXF32)

But agree its good to release this one. Looks good like this.
One small thing I noticed (not sure its a local problem)
When you click the notification, it doesnt scroll to the correct step (but the step-comments are opened)
Video below

https://user-images.githubusercontent.com/13672737/215324180-bece90a7-ff29-4027-8d25-26a6ee0f72a3.mov

thisislawatts commented:
Good catch @davehakkens, I had adjusted the way we declare addresses for comments in Research articles. I've added a test to explicitly capture this behaviour. Please pull down and generate a _new_ mention to test the behaviour. Any old notifications created locally will not work. This only impacts people who have been reviewing this branch.

davehakkens commented:
works now!

chrismclarke commented on packages/cypress/src/integration/research/comments.spec.ts:
```diff
@@ -3,13 +3,21 @@ describe('[Research]', () => {
-    cy.get('[data-cy="comment"]').should('have.length.gte', 1)
-    cy.get('[data-cy="comment-submit"]').should('be.disabled')
+  describe('[Open comments]', () => {
+    it.skip('using UI elements', () => {
```
nit(non-blocking)
Do you mean to be skpping these tests? If so could you add a comment to explain why and I can create a separate issue to address

chrismclarke commented on packages/cypress/src/support/custom-assertions.ts:
```diff
@@ -14,6 +14,23 @@ declare global {
 }
 
+chai.use((chaiObj, utils) => {
+  function assertIsInViewport() {
```
```suggestion
/**
   * Assert if the element chained from a command is in the viewport
   * @example
   * ```
   * cy.get('h1').should('be.inViewport');
   * cy.get('h1').should('not.be.inViewport');
   * ```
   * Adapted from https://www.webtips.dev/webtips/cypress/check-if-element-is-in-viewport
   */
  function assertIsInViewport() {
``````suggestion
  function assertIsInViewport() {
```

chrismclarke commented on src/pages/Research/Content/ResearchArticle.tsx:
```diff
@@ -23,6 +23,20 @@ import ReactGA from 'react-ga4'
+function areCommentVisible(updateIndex) {
+  let showComments = false
+
+  if (new RegExp(/#update-\d-comment/).test(window.location.hash)) {
```
nit(blocking)
This regex/script is used in quite a few places so might be better extracted into a utility function somehow.

Additionally, I'm not sure if we have other design limitations elsewhere, but the regex wouldn't match double-digit numbers in the url (not sure if these are possible?). e.g. #update-10-commenet
Might be better as `/#update-\d+-comment/` throughout

thisislawatts replied:
Good catch, let me take a look at refactoring this.

thisislawatts replied:
Updated in: 8a4df923dd36ac2874e29ccd515a735c7df896cb

chrismclarke commented on src/pages/common/Header/Header.tsx:
```diff
@@ -101,7 +101,7 @@ export class Header extends Component {
     const user = this.injected.userNotificationsStore.user
     const notifications = getFormattedNotifications(
-      this.injected.userNotificationsStore,
+      this.injected.userNotificationsStore.getUnreadNotifications(),
```
nicer passing in the list instead of the store for sure, thanks for the change

chrismclarke commented on functions/src/emulator/seed/content-generate.ts:
```diff
@@ -70,6 +70,7 @@ async function setMockNotifications(user: IMockAuthUser) {
         howto_mention: true,
         new_comment_research: true,
         research_useful: true,
+        research_mention: true,
```
appreciate the systematic approach to naming for research module... just wish we had thought of that for howtos!

chrismclarke commented on packages/cypress/src/support/custom-assertions.ts:
```diff
@@ -14,6 +14,23 @@ declare global {
+    ).to.eq(true)
+  }
+
+  chaiObj.Assertion.addMethod('inViewport', assertIsInViewport)
```
(Also nice to have, thanks!)

chrismclarke commented on src/pages/common/Header/getFormattedNotifications.test.tsx:
```diff
@@ -0,0 +1,19 @@
+        FactoryNotification({ type }),
+      ])
+      const { container } = render(
+        <MemoryRouter>{notification.children}</MemoryRouter>,
```
cool, never used the MemoryRouter before, can see it being useful for testing

chrismclarke commented on src/stores/Research/research.store.test.ts:
```diff
@@ -283,10 +289,45 @@ describe('research.store', () => {
         )
       })
+
+      it('triggers a notification when editing an comment to add @mention', async () => {
```
Thanks for adding all the tests to make clear the behaviour. I'll be interested to see how this works in practice, I know platforms like slack (and maybe github?) don't trigger new notifications on edited comments, but I think for us it makes sense (can't imagine people frequently editing comments, and can always find a better way to resolve multiple edits when sending out things like email notifications)

chrismclarke reviewed (changes requested):
Thanks @thisislawatts , there's quite a bit here but it all makes sense to me so thanks for the addition.
I've added one (semi) blocking nit related to the regex used to check whether the page url references mentions, and then just a couple other minor bits of code tidying.

Let me know if you think you will get a chance to review these in the next week or so, if not I can just include my suggested fix for the regex and do any other tidying at some point in the future.

chrismclarke reviewed (approved):
Thanks for all the changes @thisislawatts, looks good to go to me!
